src/coap_mbedtls.c

```diff
@@ -0,0 +1,1825 @@
+    ret = do_mbedtls_handshake(c_session, m_env);
+    if (ret == 1) {
+      /* Just connected, so send the data */
+      return coap_dtls_send(c_session, data, data_len);
+    }
+    ret = -1;
+  }
+
+  if (c_session->dtls_event >= 0) {
+    /* COAP_EVENT_DTLS_CLOSED event reported in coap_session_disconnected() */
+    if (c_session->dtls_event != COAP_EVENT_DTLS_CLOSED)
+      coap_handle_event(c_session->context, c_session->dtls_event, c_session);
+    if (c_session->dtls_event == COAP_EVENT_DTLS_ERROR ||
+      c_session->dtls_event == COAP_EVENT_DTLS_CLOSED) {
+      coap_session_disconnected(c_session, COAP_NACK_TLS_FAILED);
+      ret = -1;
+    }
+  }
+  return ret;
+}
+
+int coap_dtls_is_context_timeout(void)
+{
+  return 0;
+}
+
+coap_tick_t coap_dtls_get_context_timeout(void *dtls_context UNUSED)
+{
+  return 0;
+}
+
+coap_tick_t coap_dtls_get_timeout(coap_session_t *c_session, coap_tick_t now)
+{
+  coap_mbedtls_env_t *m_env = (coap_mbedtls_env_t *)c_session->tls;
+  int ret = mbedtls_timing_get_delay(&m_env->timer);
+
+  switch (ret) {
+  case 0:
+  case 1:
+    /* int_ms has timed out, but not fin_ms */
+    return now + 1;
+  case 2:
+    /* fin_ms has timed out - time for a retry */
+    return now;
+  default:
+    break;
+  }
+
+  return 0;
+}
+
+void coap_dtls_handle_timeout(coap_session_t *c_session)
+{
+  coap_mbedtls_env_t *m_env = (coap_mbedtls_env_t *)c_session->tls;
+
+  assert(m_env != NULL);
+  if (((c_session->state == COAP_SESSION_STATE_HANDSHAKE) &&
+       (++c_session->dtls_timeout_count > c_session->max_retransmit)) ||
+      (do_mbedtls_handshake(c_session, m_env) < 0)) {
+    /* Too many retries */
+    coap_session_disconnected(c_session, COAP_NACK_TLS_FAILED);
+  }
+  return;
+}
+
+int coap_dtls_receive(coap_session_t *c_session,
+                      const uint8_t *data,
+                      size_t data_len)
+{
+  int ret = 1;
+
+  c_session->dtls_event = -1;
+  coap_mbedtls_env_t *m_env = (coap_mbedtls_env_t *)c_session->tls;
+  assert(m_env != NULL);
+
+  coap_ssl_t *ssl_data = &m_env->coap_ssl_data;
+  if (ssl_data->pdu_len) {
+    coap_log(LOG_INFO, "** %s: Previous data not read %u bytes\n",
+             coap_session_str(c_session), ssl_data->pdu_len);
+  }
+  ssl_data->pdu = data;
+  ssl_data->pdu_len = (unsigned)data_len;
+
+  if (m_env->established) {
+#if COAP_CONSTRAINED_STACK
+    static coap_mutex_t b_static_mutex = COAP_MUTEX_INITIALIZER;
+    static uint8_t pdu[COAP_RXBUFFER_SIZE];
+#else /* ! COAP_CONSTRAINED_STACK */
+    uint8_t pdu[COAP_RXBUFFER_SIZE];
+#endif /* ! COAP_CONSTRAINED_STACK */
+
+#if COAP_CONSTRAINED_STACK
+    coap_mutex_lock(&b_static_mutex);
+#endif /* COAP_CONSTRAINED_STACK */
+
+    if (c_session->state == COAP_SESSION_STATE_HANDSHAKE) {
+      coap_handle_event(c_session->context, COAP_EVENT_DTLS_CONNECTED,
+                        c_session);
+      coap_session_connected(c_session);
+    }
+
+    ret = mbedtls_ssl_read(&m_env->ssl, pdu, (int)sizeof(pdu));
```

Comment:
`mbedtls_ssl_read()` takes a `size_t` as third argument hence it makes no sense to cast the result from `sizeof` into `int`.

Reply:
Corrected